src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      for (const [key, value] of Object.entries(action.extraData)) {
+        newState[key] = value;
+      }
+      history.push({ ...newState });
```

Comment:
The action of pushing the state copy into the history array is repeated in every case. Consider performing this action once after the switch statement to adhere to the DRY (Don't Repeat Yourself) principle.